### PR TITLE
Fix Issue 18584 - Undefined identifier when not specifying 'this'

### PR DESCRIPTION
**Pull request by RazvanN7**
Alternate fix to : #7991

### CODE REVIEW

**Comment by andralex** [comment]
Cool! cc @JinShil @WalterBright

**Comment by RazvanN7** [comment]
Auto-tester failure seems unrelated

File: src/dmd/dscope.d [hunk 1]
```diff
@@ -390,7 +390,7 @@ struct Scope
 
             static void printMsg(string txt, Dsymbol s)
             {
-                printf("%.*s  %s.%s, kind = '%s'\n", cast(int)msg.length, msg.ptr,
+                printf("%.*s  %s.%s, kind = '%s'\n", cast(int)txt.length, txt.ptr,
```
[the remark this answers is not in the document]
Comment (RazvanN7) [reply src/dmd/dscope.d]: I assume that `msg` was renamed to `txt` and it was forgotten to update the printf. I stumbled upon it while solving the issue and I made the modification.

**Comment by JinShil** [comment]
For the test case:
```D
struct S
{
    int n;
    auto fun()
    {
        return tmp!(a => n)();
    }
}

struct tmp(alias fns)
{
    alias fun = fns!int;
}
```

Scope search before:
```
Scope.search(0x7fbd9d6316c0, 'n' flags=x0)
	scope __anonymous
	scope __anonymous
	scope tmp
	scope tmp!((a) => n)
	scope __anonymous
	scope main
	scope __anonymous
```
Scope search after:
```
Scope.search(0x7f2259fbb790, 'n' flags=x0)
	scope __anonymous
	scope __anonymous
	scope fns!int
	scope __anonymous
	scope __anonymous
	scope __anonymous
	scope S                          <---- GOOD
	scope main
	scope __anonymous
```

Seeing `scope S` in the scope search is good, but I'm afraid I don't have enough understanding of the semantics well enough to certify if the rest is correct or not (hence why I was unable to fix this issue myself).  I defer judgement to those more senior than I.

**Review by andralex — APPROVED** [review]

File: src/dmd/dsymbolsem.d [hunk 2]
```diff
@@ -5591,12 +5590,20 @@ void templateInstanceSemantic(TemplateInstance tempinst, Scope* sc, Expressions*
     else if (tempinst.tinst)
     {
         bool doSemantic3 = false;
-        if (sc.func && tempinst.aliasdecl && tempinst.aliasdecl.toAlias().isFuncDeclaration())
+        FuncDeclaration fd;
+        if (tempinst.aliasdecl)
+            fd = tempinst.aliasdecl.toAlias().isFuncDeclaration();
```
Comment (stefan-koch-sociomantic) [inline src/dmd/dsymbolsem.d]: use toAlias2

**Review by stefan-koch-sociomantic — CHANGES_REQUESTED** [review]
Please use toAlias2

**Pull request merged**